File: LeetCode/l222.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def ck(self,mid:int,d:int,root:TreeNode)->bool:
        l,r=0,2**d - 1
        for i in range(d):
            t=l+(r-l)//2
            if mid<=t:
                root=root.left
                r=t
            else:
                root=root.right
                l=t+1
        return root is not None
    def countNodes(self, root: TreeNode) -> int:
        if root==None:
            return 0
        dep=0
        tmp=root
        while tmp.left!=None:
            dep+=1
            tmp=tmp.left
        l,r=1,2**dep - 1
        while l<=r:
            mid=l+(r-l)//2
            if(self.ck(mid,dep,root)):
                l=mid+1
            else:
                r=mid-1
        return (2**dep - 1) + l
        

File: LeetCode/l222.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int h=0,p,stk[40];
    bool ck(int x,TreeNode * root){
        p=0;x--;
        for(int i=1;i<=h;i++){
            stk[i]=x&1;
            x>>=1;
        }

        for(int i=h;i>=1; i--){
            if(stk[i]==0){
                if(root->left){
                    root=root->left;
                }else return false;
            }else{
                if(root->right){
                    root=root->right;
                }else return false;
            }
        }
        return true;

        
    }
    int countNodes(TreeNode* root) {
        if(!root)return 0;
        auto u=root;
        while(u->left)u=u->left,h++;
        int l=1,r=1<<h;
        while(l<r){
            int mid=l+(r-l+1)/2;
            if(ck(mid,root)){
                l=mid;
            }else r=mid-1;
        }
        return l+(1<<h)-1;
    }
};